src/tests/end2end/CopyTests.cpp

```diff
@@ -0,0 +1,418 @@
+// Copyright 2017 The NXT Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "tests/NXTTest.h"
+
+#include <array>
+#include "common/Constants.h"
+#include "common/Math.h"
+#include "utils/NXTHelpers.h"
+
+class CopyTests : public NXTTest {
+    protected:
+        static constexpr unsigned int kBytesPerTexel = 4;
+
+        struct SourceTexture {
```

Comment:
nit: SourceTexture could be renamed to account for B->T copies done in this test.